README.md
# 自然布局

<p align="center">
  <a href="https://fusion.design/">
    <img alt="Fusion" src="https://img.alicdn.com/tfs/TB1YsoiHVzqK1RjSZFCXXbbxVXa-159-99.svg" width="200">
  </a>
</p>
<p align="center">
  <img src="https://img.shields.io/npm/v/@alifd/layout.svg?style=flat-square">
  <img src="https://img.shields.io/npm/dm/@alifd/layout?style=flat-square">
  <img src="https://img.shields.io/npm/l/@alifd/layout.svg?style=flat-square">
</p>

用于快速完成页面布局体系 (低代码搭建或源码开发)，目标是不写任何布局 CSS 代码。

**设计理念**

- **初衷**：愿布局不需写 css 样式
- **定位**：页面内容区的布局解决方案

> 详细介绍：https://www.yuque.com/fusion-design/layout/yzx8g4 （需先加入 [Fusion Design](https://www.yuque.com/g/fusion-design/layout/collaborator/join?token=7bTjikyxDTAh3WwS) 语雀知识库）

预览

- [预览 Demo](https://g.alipay.com/@alifd/layout/build/index.html)
- [阿里低代码搭建 Demo](https://g.alipay.com/@alifd/layout/build/lowcode/index.html)

## 安装

```bash
 npm install @alifd/layout --save
```

## 使用

在项目中，引入一套 Fusion Design 主题的 [Design Token](https://fusion.design/pc/design-tokens), 如：

```jsx
import '@alifd/theme-3/variables.css';
```

> 注：此处是为了引入布局组件所必须得 CSS Variables, 只需要保证上述的 CSS 代码生效即可，引入方式不限。

布局示例：

```jsx
import { Page, Section, Block, Row, Col, Cell, P, Text, Space } from '@alifd/layout';

export default function App() {
  return (
    <Page>
      <Page.Header>Header</Page.Header>
      <Page.Content>
        <Section>
          <Block>
            <Row>
              <Cell>
                <P>
                  <Text>文本</Text>
                  <Text>文本</Text>
                </P>
              </Cell>
              <Cell>
                <P>
                  <Text>文本</Text>
                  <Text>文本</Text>
                </P>
              </Cell>
            </Row>
            <Space />
            <Col>
              <Cell>
                <P>
                  <Text>文本</Text>
                  <Text>文本</Text>
                </P>
              </Cell>
              <Cell>
                <P>
                  <Text>文本</Text>
                  <Text>文本</Text>
                </P>
              </Cell>
            </Col>
          </Block>
        </Section>
      </Page.Content>
      <Page.Footer>Footer</Page.Footer>
    </Page>
  );
}
```

### 进一步封装 Block 和 Text。

由于 `Section > Block` 具有强关联性（依赖 `Block` 对子元素宽度做计算并调整）。如果业务上需要进一步封装 `Block`，需要对 `Block` 增加类型标记，一遍 Section 能识别其 `span` 等属性。

同理 `P > Text` 也具有类似强相关性，对于 `Text` 的组件封装，如果希望在 `P` 中表现和 `Text` 近似，也需要标记 `typeMark`.

示例：

```jsx
import { Page, Section, Block } from '@alifd/layout';

function NewBlock(props) {
  const { children, ...others } = props;
  // 加入新业组件逻辑
  return <Block {...others}>{children}</Block>;
}

// 标记 NewBlock 可以视为 Block 作为 Section 的子元素
NewBlock.typeMark = 'Block';

// 页面渲染可以使用 NewBlock
function App() {
  return (
    <Page>
      <Section>
        <Block span={2}>block</Block>
        <NewBlock span={10}>new block</NewBlock>
      </Section>
    </Page>
  );
}
```

## 开发

1. 调试和预览 demo

```bash
npm run start
```

2. 启动低代码预览

```bash
npm run lowcode:dev
```

## API

### Page

页面

| 参数               | 含义                                        | 类型                                              | 默认值       |
| ------------------ | ------------------------------------------- | ------------------------------------------------- | ------------ |
| prefix             | CSS 类名前缀                                | string                                            | `fd-layout-` |
| minHeight          | 页面的最小高度，例如 `calc(100vh - 52px)`   | Number/String                                     | -            |
| noPadding          | 禁用页面内边距,包含 Header, Content, Footer | Boolean                                           | `false`      |
| sectionGap         | Header、Footer、Nav、Aside 和章节之间间隙   | Number                                            | -            |
| blockGap           | Section 中栅格布局间隙                      | Number                                            | -            |
| gridGap            | 小布局间隙（包含行、列、网格布局）          | Number                                            | -            |
| breakPoints        | 断点信息                                    | `BreakPoint[]`                                    | -            |
| children           | 子元素                                      | ReactNode                                         | -            |
| onBreakPointChange | 断点变更回调                                | (curBreakPoint, oldBreakPoint, breakPoints)=>void |              |

```ts
interface BreakPoint {
  /**
   *  断点宽度(包含)
   */
  width: number;
  /**
   * 最大展示宽, 默认为断点宽度
   */
  maxContentWidth: number | string;
  /**
   * 列数
   */
  numberOfColumns: 1 | 2 | 4 | 8 | 12 | 16 | 24;
}
```

#### Page.Header

序章

| 参数            | 含义                                          | 类型      | 默认值 |
| --------------- | --------------------------------------------- | --------- | ------ |
| mode            | 背景色 'lining' / 'surface' / 'transparent'   | Enum      | -      |
| noPadding       | 移除内边距                                    | Boolean   | false  |
| noBottomPadding | 隐藏默认底部内边距                            | Boolean   | false  |
| noBottomMargin  | 隐藏与内容部分的间隙(外边距)                  | Boolean   | false  |
| divider         | 展示分割线                                    | Boolean   | -      |
| fullWidth       | 保持全屏宽（不受断点 `maxContentWidth` 限制） | Boolean   | false  |
| children        | 子元素                                        | ReactNode | -      |

#### Page.Footer

终章

| 参数         | 含义                                          | 类型      | 默认值 |
| ------------ | --------------------------------------------- | --------- | ------ |
| mode         | 背景色 'lining' / 'surface' / 'transparent'   | Enum      | -      |
| noTopPadding | 隐藏默认顶部内边距                            | Boolean   | -      |
| divider      | 展示分割线                                    | Boolean   | -      |
| fullWidth    | 保持全屏宽（不受断点 `maxContentWidth` 限制） | Boolean   | false  |
| children     | 子元素                                        | ReactNode | -      |

#### Page.Content

内容

| 参数      | 含义                                    | 类型          | 默认值 |
| --------- | --------------------------------------- | ------------- | ------ |
| minHeight | 页面的最小高度，例如 calc(100vh - 52px) | Number/String | -      |
| children  | 子元素                                  | ReactNode     | -      |

#### Page.Aside

右侧边栏，如有 Content，需为 Content 的子元素

| 参数     | 含义   | 类型          | 默认值 |
| -------- | ------ | ------------- | ------ |
| width    | 宽度   | Number/String | -      |
| children | 子元素 | ReactNode     | -      |

#### Page.Nav

左侧导航，如有 Content，需为 Content 的子元素

| 参数     | 含义   | 类型          | 默认值 |
| -------- | ------ | ------------- | ------ |
| width    | 宽度   | Number/String | -      |
| children | 子元素 | ReactNode     | -      |

### Section

章

| 参数       | 含义                          | 类型      | 默认值 |
| ---------- | ----------------------------- | --------- | ------ |
| title      | 标题                          | ReactNode | -      |
| titleAlign | 标题位置，可选值 left/center  | Enum      | -      |
| extra      | 附加区域（标题右侧）          | ReactNode | -      |
| gap        | 自定义内部区块（Block）的间隙 | Number    | -      |
| children   | 子元素                        | ReactNode | -      |

### Block

区块（节）

| 参数             | 含义                                                              | 类型          | 默认值 |
| ---------------- | ----------------------------------------------------------------- | ------------- | ------ |
| mode             | 背景模式，可选值 transparent(自动移除内边距和标题)/surface/lining | Enum          | -      |
| title            | 标题                                                              | ReactNode     | -      |
| titleAlign       | 标题位置，可选值 left/center                                      | Enum          | -      |
| extra            | 附加区域（标题右侧）                                              | ReactNode     | -      |
| divider          | 展示标题与内容之间的分割线                                        | Boolean       | -      |
| bordered         | 展示边框                                                          | Boolean       | -      |
| noPadding        | 移除内边距                                                        | Boolean       | false  |
| span             | 列宽                                                              | Number        | -      |
| width            | 指定宽度                                                          | Number        | -      |
| contentClassName | 有标题时，内容区域的样式名                                        | string        | -      |
| contentStyle     | 有标题时，内容区域的自定义样式                                    | CSSProperties | -      |
| children         | 子元素                                                            | ReactNode     | -      |

### Row

行

| 参数     | 含义                                                      | 类型          | 默认值 |
| -------- | --------------------------------------------------------- | ------------- | ------ |
| width    | 指定宽度                                                  | Number/String | -      |
| height   | 指定高度                                                  | Number/String | -      |
| autoFit  | 根据内容自适应宽度（当作为行列布局的子元素时生效）        | Boolean       | -      |
| gap      | 自定义元素间间距                                          | Number        | -      |
| verAlign | 垂直对齐方式， 可选值：top/middle/bottom/stretch/baseline | Enum          | -      |
| children | 子元素                                                    | ReactNode     | -      |

### Col

列

| 参数     | 含义                                               | 类型          | 默认值 |
| -------- | -------------------------------------------------- | ------------- | ------ |
| autoFit  | 根据内容自适应宽度（当作为行列布局的子元素时生效） | Boolean       | -      |
| width    | 指定宽度                                           | Number/String | -      |
| height   | 指定高度                                           | Number/String | -      |
| gap      | 自定义元素间间距                                   | Number        | -      |
| align    | 水平对齐方式， 可选值：left/center/right/stretch   | Enum          | -      |
| children | 子元素                                             | ReactNode     | -      |

### Grid

网格

| 参数       | 含义                                          | 类型                         | 默认值 |
| ---------- | --------------------------------------------- | ---------------------------- | ------ |
| cols       | 指定列数                                      | number                       |        |
| rows       | 指定行数                                      | number                       |        |
| minWidth   | 单元格最小宽度                                | number                       |        |
| maxWidth   | 单元格最大宽度                                | number                       |        |
| rowGap     | 单元格行间距, 可选值: small/medium/large/数值 | Enum                         |        |
| colGap     | 单元格列间距, 可选值: small/medium/large/数值 | Enum                         |        |
| renderItem | 手动渲染单个单元格内容                        | (rowIndex,colIndex)=>RaxNode |        |
| children   | 子元素，默认应为 `Cell`                       | RaxNode                      | -      |

### Cell

单元格，其内容默认为 flex 纵向布局

| 参数     | 含义                                                                                 | 类型          | 默认值 |
| -------- | ------------------------------------------------------------------------------------ | ------------- | ------ |
| width    | 指定宽度                                                                             | Number/String | -      |
| height   | 指定高度                                                                             | Number/String | -      |
| autoFit  | 根据内容自适应宽度（当作为行列布局的子元素时生效）                                   | Boolean       | -      |
| gap      | 自定义内部元素的行解析                                                               | Number        | 0      |
| align    | 内容水平对齐方式， 可选值: left/center/right                                         | Enum          | -      |
| verAlign | 内容垂直对齐方式， 可选值: top/middle/bottom/space-between/space-around/space-evenly | Enum          | -      |
| block    | 使用 `block` 布局                                                                    | Boolean       | -      |
| children | 子元素                                                                               | ReactNode     | -      |

### Space

空间间隙

| 参数      | 说明                                     | 类型      | 默认值 |
| --------- | ---------------------------------------- | --------- | ------ |
| direction | 组件自身布局模式，可选： `hoz/ver`       | Enum      | hoz    |
| size      | 尺寸， 可选: `small/medium/large/Number` | Enum      | medium |
| children  | 子元素                                   | ReactNode |        |

### 文本

| 参数      | 说明                                     | 类型    | 默认值 |
| --------- | ---------------------------------------- | ------- | ------ |
| direction | 组件自身布局模式，可选： `hoz/ver`       | Enum    | hoz    |
| size      | 尺寸， 可选: `small/medium/large/Number` | Enum    | medium |
| children  | 子元素                                   | RaxNode |        |

### P

段落

| 参数          | 说明                                                                       | 类型      | 默认值     |
| ------------- | -------------------------------------------------------------------------- | --------- | ---------- |
| align         | 水平方向对齐模式 left/center/right/space-between/space-around/space-evenly | Enum      | 'left'     |
| verAlign      | 垂直方向对齐模式 top/middle/bottom/baseline                                | Enum      | 'baseline' |
| spacing       | 子元素间保持水平间距, 可选： small/medium/large/false                      | Enum      | medium     |
| hasVerSpacing | 除 `文本` 节点外子元素间保持垂直间距                                       | Boolean   | true       |
| beforeMargin  | 段前外边距（作为第一个子元素时无效）                                       | Number    | 0          |
| afterMargin   | 段后外边距（作为最后一个子元素时无效）                                     | Number    | 0          |
| children      | 子元素                                                                     | ReactNode | -          |

#### Text

文本

| 参数      | 说明                                                              | 类型    | 默认值 |
| --------- | ----------------------------------------------------------------- | ------- | ------ |
| type      | 约束字体大小 overline/caption/body1/body2/title/h1/h2/h3/h4/h5/h6 | Enum    | body2  |
| delete    | 添加删除线样式                                                    | Boolean | false  |
| mark      | 添加标记样式                                                      | Boolean | false  |
| underline | 添加下划线样式                                                    | Boolean | false  |
| strong    | 是否加粗                                                          | Boolean | false  |
| code      | 添加代码样式                                                      | Boolean | false  |
| component | 设置标签类型                                                      | custom  | 'span' |
| color     | 颜色                                                              | String  | -      |

## CSS 变量

自然布局默认引入了 Fusion 设计系统的 [Design Tokens](https://fusion.design/pc/design-tokens?type=theme&themeid=3)。
布局可自定义 `CSS Variables` 如下，可基于实际情况修改：

| 变量名                            | 说明                   | 默认值                 |
| --------------------------------- | ---------------------- | ---------------------- |
| `--color-transparent`             | 前景色                 | `transparent`          |
| `--color-surface`                 | 前景色                 | `#fff`                 |
| `--color-lining`                  | 衬色                   | `#f0f0f0`              |
| `--page-padding-lr`               | 页面左右内边距         | `var(--s-5)`           |
| `--page-padding-tb`               | 页面上下内边距         | `var(--s-5)`           |
| `--page-header-divider-color`     | header 分割线颜色      | `var(--color-line1-1)` |
| `--page-block-corner`             | 区块圆角尺寸           | `var(--corner-0)`      |
| `--page-block-padding-lr`         | 区块左右内边距         | `var(--s-3)`           |
| `--page-block-padding-tb`         | 区块上下内边距         | `var(--s-3)`           |
| `--page-section-title-font-color` | 章节标题颜色           | `var(--color-text1-4)` |
| `--page-section-extra-font-color` | 章节附加内容文本颜色   | `var(--color-text1-2)` |
| `--page-block-title-font-color`   | 区块标题颜色           | `var(--color-text1-4)` |
| `--page-block-extra-font-color`   | 区块附加内容文本颜色   | `var(--color-text1-2)` |
| `--page-section-gap`              | 章间隙                 | `var(--s-3)`           |
| `--page-block-gap`                | 区块间隙               | `var(--s-2)`           |
| `--page-block-border-width`       | 区块边框宽度           | `var(--line-1)`        |
| `--page-block-border-color`       | 区块边框颜色           | `var(--color-line1-1)` |
| `--page-grid-gap`                 | 小布局间隙             | `var(--s-1)`           |
| `--page-p-small-spacing`          | 段落子元素的水平小间距 | `var(--s-1)`           |
| `--page-p-medium-spacing`         | 段落子元素的水平中间距 | `var(--s-2)`           |
| `--page-p-large-spacing`          | 段落子元素的水平大间距 | `var(--s-4)`           |
| `--page-p-el-margin`              | 段落子元素的上下间距   | `var(--s-1)`           |
| `--page-p-font-color`             | 段落默认字体色         | `var(--color-text1-4)` |


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [2.4.1](https://github.com/alibaba-fusion/layout/compare/v2.4.0...v2.4.1) (2023-10-12)


### Bug Fixes

* 修复自由布局错误和不能拖拽的问题 ([a4c7576](https://github.com/alibaba-fusion/layout/commit/a4c757627ac417523a725b1fa331eed7797ff533))

## [2.4.0](https://github.com/alibaba-fusion/layout/compare/v2.3.0...v2.4.0) (2023-08-30)


### Features

* 去除自然布局的 disableBehaviors, 使其可以被删除 ([798418c](https://github.com/alibaba-fusion/layout/commit/798418cff9fb4876d626b7cc3de0df5aa4d496cf))

## [2.3.0](https://github.com/alibaba-fusion/layout/compare/v2.2.4...v2.3.0) (2023-08-01)


### Features

* 补齐css 变量默认值 ([75ccc12](https://github.com/alibaba-fusion/layout/commit/75ccc129c6a5db8684a6ee9e69ba572225df0575))

### [2.2.4](https://github.com/alibaba-fusion/layout/compare/v2.2.3...v2.2.4) (2023-07-28)

### [2.2.3](https://github.com/alibaba-fusion/layout/compare/v2.2.2...v2.2.3) (2023-07-28)

### [2.2.2](https://github.com/alibaba-fusion/layout/compare/v2.2.1...v2.2.2) (2023-07-27)


### Bug Fixes

* lowcode_es/view 内 src 引用报错 ([1f33a1f](https://github.com/alibaba-fusion/layout/commit/1f33a1f1ef7e0dda6c627235f289c66f98e96d1d))

### [2.2.1](https://github.com/alibaba-fusion/layout/compare/v2.2.0...v2.2.1) (2023-07-27)

## 2.2.0 (2023-07-27)


### Features

* 调整 peer ([594f5ef](https://github.com/alibaba-fusion/layout/commit/594f5ef1cb7ccc462f29098ffeb6427052fe8a8f))
* 调整使用文档链接 ([5b3dfca](https://github.com/alibaba-fusion/layout/commit/5b3dfcacbac7cdd1e4d038a336a60f6b650b0093))
* 段落 verMargin 更名为 hasVerSpacing ([6484ee8](https://github.com/alibaba-fusion/layout/commit/6484ee82026ce31f4459ea2542b4b130da79c5f0))
* 上传新的开源版本 ([2f3af5c](https://github.com/alibaba-fusion/layout/commit/2f3af5cd6e5cd6b13a3ed3b6abfe224459a3a294))
* 升级 build-plugin-lowcode，适配 lowcode 打包 ([d82a6c6](https://github.com/alibaba-fusion/layout/commit/d82a6c663a8e0dd7d7f64417243e9eca8f7d86f7))
* 使用 lodash-es ([42172a3](https://github.com/alibaba-fusion/layout/commit/42172a35cf93de4c8e42338ad1a379d5cbd48dbd))
* 移除 tab 功能 ([e7a21ed](https://github.com/alibaba-fusion/layout/commit/e7a21ed3818adcad9dc9b4b7b26e3ddd17b3813f))
* 移除低代码分页配置 ([5b4efc4](https://github.com/alibaba-fusion/layout/commit/5b4efc4395278593820f940786d5b4a9e43a52d1))
* 移除x-if 判断 ([0c20e48](https://github.com/alibaba-fusion/layout/commit/0c20e48a2806c30ccc21fb3fe736636113ad85fa))
* 优化部分 section 和 page 的代码 ([6c9cbc9](https://github.com/alibaba-fusion/layout/commit/6c9cbc98d218a7ef1881e9d33df651bba7017db9))
* 自动引入 css ([f0d2b08](https://github.com/alibaba-fusion/layout/commit/f0d2b08997ca68127dcdd80ff4c553dcae09ebf9))
* 最大支持 24 栅栏 ([a0544de](https://github.com/alibaba-fusion/layout/commit/a0544de80c82442e526d97f5d6bf72d01366b75f))
* add adaptors for design components ([1526f4a](https://github.com/alibaba-fusion/layout/commit/1526f4a9ccb596eaf73ca1f356587bb426d50904))
* add demo ([3c2907f](https://github.com/alibaba-fusion/layout/commit/3c2907f34fb689fcb03240c7f75c7bd04b50ecc0))
* add tabletColSpan phoneColSpan for responsive ([431712c](https://github.com/alibaba-fusion/layout/commit/431712c9af349f17b5fd12f9bda703fccc7831cf))
* add ts ([40a305e](https://github.com/alibaba-fusion/layout/commit/40a305e321645f12a80be06833a68843b8279f92))
* add types for AliLowcodeEngine ([2ba5718](https://github.com/alibaba-fusion/layout/commit/2ba57186a916e43b2e1135713889f5ea23b18f26))
* alias import ([9b92166](https://github.com/alibaba-fusion/layout/commit/9b921665c80ddddefe1d1736392586ac278004a4))
* fix conflict ([969a3c5](https://github.com/alibaba-fusion/layout/commit/969a3c5723edec0bc33c4283b5710dfea811f1f2))
* fix conflict ([888a3ea](https://github.com/alibaba-fusion/layout/commit/888a3ea9d98d9d95aae1b487296df970b896d15b))
* lowcode ([c4a3f99](https://github.com/alibaba-fusion/layout/commit/c4a3f9959c66b3f5f10b15869769559c5a6ecc40))
* p, space 支持数值类型的 spacing 和 size 传入 ([d629c79](https://github.com/alibaba-fusion/layout/commit/d629c79445f5cc3b3ef160d07595213c48cfe0ab))
* Row/Col/Cell 支持height ([f55273a](https://github.com/alibaba-fusion/layout/commit/f55273a9741263bdf3dd30c3f87e20065a97a394))
* suport lowcodeEngine 2.0 ([33d9299](https://github.com/alibaba-fusion/layout/commit/33d9299837928f65c303ebe155157213cfae3806))
* update lock.json ([5057bc7](https://github.com/alibaba-fusion/layout/commit/5057bc70e8dfe27132cb602b2c1aa0f43fcf19cc))
* update version of lowcode-engine and lowcode-materials ([087c58a](https://github.com/alibaba-fusion/layout/commit/087c58a510c5c61863ede54f8b55c08337940784))


### Bug Fixes

*  header 不支持 noBottomPadding ([255c115](https://github.com/alibaba-fusion/layout/commit/255c1154b6d4eda957106828c7c9d1749ca832e8))
* 部分场景下布局渲染异常 ([4831f6d](https://github.com/alibaba-fusion/layout/commit/4831f6d4d3e776a87a63b6df203cd081e3f71264))
* 低代码引擎升级导致的容器组件内小剪刀无法使用 ([7460625](https://github.com/alibaba-fusion/layout/commit/7460625bab4d45f7113f8e9c6d082542c2f51196))
* 断点调整错误 ([768beb4](https://github.com/alibaba-fusion/layout/commit/768beb48c302388a02dea3cd527214e98a6e3ac8))
* 构建产物 types 路径无效 ([0971d41](https://github.com/alibaba-fusion/layout/commit/0971d41599df8290ba321db6cfa011bd9fe5ccc9))
* 类名错误 ([d4f1e4c](https://github.com/alibaba-fusion/layout/commit/d4f1e4c4977657c4e7c459f5902fa403687ed613))
* 拖拽修改容器高度，导致容器变空白 ([04674ed](https://github.com/alibaba-fusion/layout/commit/04674ed43cb0692e390dfdd3418d6bf77f116c46))
* 拖拽修改容器高度，导致容器变空白 ([d5383fe](https://github.com/alibaba-fusion/layout/commit/d5383fe65a9b53c85b34b176267b8b54b254c3b8))
* 选中容器时，布局页面会闪一下 ([1e52924](https://github.com/alibaba-fusion/layout/commit/1e52924f58a7bc1dcdb98de4299dce265b574bbe))
* 自由容器 flex 布局 ([9a4e642](https://github.com/alibaba-fusion/layout/commit/9a4e642da44e1f2d1ddf53fccae846c7619edf44))
* add callback ([d06e895](https://github.com/alibaba-fusion/layout/commit/d06e895e989c569b20c11f769f6a74c060a2dde1))

src/types.ts
import { CSSProperties, HTMLAttributes, ReactNode } from 'react';

export type BaseSize = 'small' | 'medium' | 'large';

/**
 * 基础属性
 */
export type BaseProps = Omit<HTMLAttributes<HTMLDivElement>, 'title'>;

/**
 * 基础的小布局简写配置
 */
export type BaseGap = 'auto' | number;

export interface BaseBgMode {
  /**
   * 容器预置背景色
   */
  mode?: 'surface' | 'lining' | 'transparent';
}

export interface TypeMark {
  typeMark?: string;
}

/**
 * 断点
 */
export interface BreakPoint {
  /**
   *  断点宽度(包含)
   */
  width: number;
  /**
   * 最大展示宽, 默认为断点宽度
   */
  maxContentWidth: number | string;
  /**
   * 列数
   */
  numberOfColumns: 1 | 2 | 4 | 8 | 12 | 16 | 24;
}

/**
 * 断点配置(至少 2 个)
 */
export type BreakPoints = BreakPoint[];

/**
 * 上下文
 */
export interface LayoutContextProps {
  /**
   * 类名前缀
   */
  prefix?: string;
  /**
   * 当前断点信息
   */
  breakPoint: BreakPoint;
  /**
   * 禁用页面内容的内边距
   */
  noPadding?: boolean;
  /**
   * 最大列数
   */
  maxNumberOfColumns: number;
  /**
   * content 中 "章" 之间的间距
   */
  sectionGap?: BaseGap;
  /**
   * 水槽间距
   */
  blockGap?: BaseGap;
  /**
   * 小布局间距（行、列、网格布局的 单元格-Cell 间距）
   */
  gridGap?: BaseGap;
}

/**
 * 内容
 */
export interface PageContentProps extends BaseProps, BaseBgMode {
  /**
   * 子节点
   */
  children?: ReactNode;
  /**
   * 禁用内容区域的默认内边距
   */
  noPadding?: boolean;
}

/**
 * 章节
 */
export interface SectionProps extends BaseProps {
  /**
   * 章节标题
   */
  title?: ReactNode;
  /**
   * 章节标题
   */
  titleAlign?: 'left' | 'center';
  /**
   * 附加内容
   */
  extra?: ReactNode;
  /**
   * 移除内边距
   */
  noPadding?: boolean;
  /**
   * 章节中区块的间隙
   */
  gap?: BaseGap;
  /**
   * 子节点
   */
  children?: ReactNode;
}

/**
 * 单元格容器
 */
export interface CellProps extends BaseProps {
  /**
   * 在行模式下，未设置 autoFit 时, 自定义单个 cell 的宽度，
   */
  width?: number | string;
  /**
   * 指定高度
   */
  height?: number | string;
  /**
   * 是否宽度（行模式下）/ 高度（列模式下）自适应内容
   */
  autoFit?: boolean;
  /**
   * 自定义内容间隙
   */
  gap?: BaseGap;
  /**
   * 主轴方向
   */
  direction?: 'hoz' | 'ver';
  /**
   * cell 子元素的水平对齐方式
   */
  align?: 'left' | 'center' | 'right';
  /**
   * cell 子元素的垂直对齐方式
   */
  verAlign?: 'top' | 'middle' | 'bottom' | 'space-around' | 'space-between' | 'space-evenly';
  /**
   * 启用 display:block 模式(默认为 flex )
   */
  block?: boolean;
  /**
   * 孩子节点
   */
  children?: ReactNode;
  [prop: string]: any;
}

/**
 * 区块，默认 flex 布局
 */
export interface BlockProps extends BaseProps, BaseBgMode, Omit<CellProps, 'width' | 'height'> {
  /**
   * 禁用 padding
   */
  noPadding?: boolean;
  /**
   * 展示标题与内容的分割线
   */
  divider?: boolean;
  /**
   * 展示边框
   */
  bordered?: boolean;
  /**
   * 列宽
   */
  span?: number;
  /**
   * 指定宽度
   */
  width?: number | string;
  /**
   * 有标题时生效,内容区域的样式名
   */
  contentClassName?: string;
  /**
   * 有标题时生效,内容区域的自定义样式
   */
  contentStyle?: CSSProperties;
  /**
   * 区块标题
   */
  title?: ReactNode;
  /**
   * 标题的对齐方式
   */
  titleAlign?: 'left' | 'center';
  /**
   * 附加内容
   */
  extra?: ReactNode;
  /**
   * 孩子节点
   */
  children?: ReactNode;
}

// ==========================小布局====================
/**
 *  cell 父容器组件
 */
export interface GridProps extends BaseProps {
  /**
   * 默认行列间距, `BaseGap`  可为预设值或 css 单位
   */
  gap?: [BaseGap, BaseGap];
  /**
   * 行数
   */
  rows?: number;
  /**
   * 列数
   */
  cols?: number;
  /**
   * 单元格的宽度
   */
  width?: number;
  /**
   * 单元格的最小宽度
   */
  minWidth?: number;
  /**
   * 单元格的最大宽度
   */
  maxWidth?: number;
  /**
   * 行间隙,可为预设值 `small`， `medium`、`large`, `false` 或 css 长度单位。此项设置会覆盖 Page 上的 grid-gap 全局值
   */
  rowGap?: BaseGap;
  /**
   * 列间隙,可为预设值 `small`， `medium`、`large`, `false` 或 css 长度单位。此项设置会覆盖 Page 上的 grid-gap 全局值
   */
  colGap?: BaseGap;
  /**
   * 水平对齐方式
   */
  align?: 'left' | 'center' | 'right' | 'stretch';
  /**
   * 垂直对齐方式
   */
  verAlign?: 'top' | 'middle' | 'bottom' | 'stretch';
  /**
   * 自定义渲染网格项
   */
  renderItem?: (row: number, col: number) => ReactNode;
  /**
   * 孩子节点
   */
  children?: ReactNode;
}

/**
 * 行列通用属性类型
 */
interface BaseRowColPropBases extends BaseProps {
  /**
   * 是否 宽度/高度 自适应
   */
  autoFit?: boolean;
  /**
   * 单元格列间距尺寸, 可为预设值 `small`， `medium`、`large`, `false` 或 css 长度单位。此项设置会覆盖 Page 上的 grid-gap 全局值
   */
  gap?: BaseGap;
  /**
   * 固定宽度
   */
  width?: number | string;
  /**
   *固定高度
   */
  height?: number | string;
  /**
   * 自定义样式
   */
  style?: CSSProperties;
  /**
   * 孩子节点
   */
  children?: ReactNode;
}

/**
 * 行模式容器
 */
export interface RowProps extends BaseRowColPropBases {
  /**
   * 子元素垂直对齐方式
   */
  verAlign?: 'top' | 'middle' | 'bottom' | 'stretch' | 'baseline';
}

/**
 * 列模式容器
 */
export interface ColProps extends BaseRowColPropBases {
  /**
   * 子元素水平对齐方式
   */
  align?: 'left' | 'center' | 'right' | 'stretch';
}

/**
 * 字体
 */
export interface TextProps extends BaseProps {
  /**
   * 类型
   */
  type?: 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6' | 'body1' | 'body2' | 'caption' | 'overline';
  /**
   * 添加删除线样式
   */
  delete?: boolean;
  /**
   * 添加标记样式
   */
  mark?: boolean;
  /**
   * 添加下划线样式
   */
  underline?: boolean;
  /**
   * 是否加粗
   */
  strong?: boolean;
  /**
   * 添加代码样式
   */
  code?: boolean;
  /**
   * 设置标签类型
   */
  component?: string;
  /**
   * 自定义颜色
   */
  color?: string;
  /**
   * 文本对齐方式
   */
  align?: 'left' | 'center' | 'right';
  /**
   * 自定义背景色
   */
  backgroundColor?: string;
  /**
   * 子节点
   */
  children?: ReactNode;
}

/**
 * 段落
 */
export interface ParagraphProps extends BaseProps {
  /**
   * 字体类型（快捷方式），相当于 Typography
   */
  type?: TextProps['type'];
  /**
   * 水平对齐方式
   */
  align?: 'left' | 'right' | 'center' | 'space-between' | 'space-around' | 'space-evenly';
  /**
   * 垂直对齐方式
   */
  verAlign?: 'top' | 'middle' | 'bottom' | 'baseline';
  /**
   * 子元素间水平间距
   */
  spacing?: BaseSize | string | number;
  /**
   * 子元素间具有垂直间距
   */
  hasVerSpacing?: boolean;
  /**
   * 段前间距尺寸（在父容器中的第一个段落无效）
   */
  beforeMargin?: number | string;
  /**
   * 段后间距尺寸 (在父容器中的最后一个段落无效)
   */
  afterMargin?: number | string;
  /**
   * 孩子节点
   */
  children?: ReactNode;
}

/**
 * 间距
 */
export interface SpaceProps extends BaseProps {
  /**
   * 尺寸
   */
  size?: BaseSize | string | number;
  /**
   * 布局方向
   */
  direction?: 'hoz' | 'ver';
  /**
   * 子元素
   */
  children?: ReactNode;
}
